### PR TITLE
[#1471] feat(server/dashboard): Add Reg time to ServerNode

### PR DESCRIPTION
**Pull request by maobaolong**
### What changes were proposed in this pull request?

Add reg time to ServerNode, and update method to update ServerNode when server heartbeat to coordinator rather than replace serverNode object.

### Why are the changes needed?

Fix: #1471

### Does this PR introduce _any_ user-facing change?

User can get registrationTime from dashboard server page.

### How was this patch tested?

image
image

### CODE REVIEW

File: coordinator/src/main/java/org/apache/uniffle/coordinator/SimpleClusterManager.java [hunk 1]
```diff
@@ -226,8 +226,11 @@ private void parseExcludeNodesFile(DataInputStream fsDataInputStream) throws IOE
   public void add(ServerNode node) {
     if (!servers.containsKey(node.getId())) {
       LOG.info("Newly registering node: {}", node.getId());
+      servers.put(node.getId(), node);
+    } else {
+      servers.get(node.getId()).update(node);
     }
```
Comment (xianjingfeng) [inline coordinator/src/main/java/org/apache/uniffle/coordinator/SimpleClusterManager.java:232]:
```suggestion
    ServerNode previousNode = servers.get(node.getId());
    if (previousNode == null) {
      LOG.info("Newly registering node: {}", node.getId());
    } else {
      node.setRegistrationTime(previousNode.getRegistrationTime());
    }
    servers.put(node.getId(), node);
```

File: coordinator/src/main/java/org/apache/uniffle/coordinator/ServerNode.java [hunk 2]
```diff
@@ -267,4 +277,18 @@ public long getTotalMemory() {
...
+    this.usedMemory = node.getUsedMemory();
+    this.preAllocatedMemory = node.getPreAllocatedMemory();
+    this.availableMemory = node.getAvailableMemory();
+    this.eventNumInFlush = node.getEventNumInFlush();
+    this.timestamp = System.currentTimeMillis();
+    this.tags = node.getTags();
+    this.status = node.getStatus();
+    this.storageInfo = node.getStorageInfo();
+    this.nettyPort = node.getNettyPort();
+  }
```
Comment (xianjingfeng) [inline coordinator/src/main/java/org/apache/uniffle/coordinator/ServerNode.java:293]:
```suggestion
```

File: coordinator/src/main/java/org/apache/uniffle/coordinator/SimpleClusterManager.java (hunk 1 above)
Comment (maobaolong) [reply coordinator/src/main/java/org/apache/uniffle/coordinator/SimpleClusterManager.java:232]: @xianjingfeng If we modified like this, the other property of ServerNode would not be updated, something like usedMemory, preAllocatedMemory, availableMemory, eventNumInFlush and so on.

**Comment by maobaolong** [comment]
@xianjingfeng Thanks for your suggestion, but I'm worried of the new ServerNode state cannot be updated, any thoughts?

**Comment by maobaolong** [comment]
@advancedxy @xianjingfeng Thanks for your suggestion, I addressed your comments, PTAL

File: coordinator/src/main/java/org/apache/uniffle/coordinator/SimpleClusterManager.java [hunk 3]
```diff
@@ -226,8 +226,13 @@ private void parseExcludeNodesFile(DataInputStream fsDataInputStream) throws IOE
   public void add(ServerNode node) {
     if (!servers.containsKey(node.getId())) {
       LOG.info("Newly registering node: {}", node.getId());
+    } else {
+      ServerNode pre = servers.get(node.getId());
+      long regTime = pre.getRegistrationTime();
+      // inherit registration time
+      node.setRegistrationTime(regTime);
     }
```
Comment (xianjingfeng) [inline coordinator/src/main/java/org/apache/uniffle/coordinator/SimpleClusterManager.java:234]:
```suggestion
    ServerNode pre = servers.get(node.getId());
    if (pre == null) {
      LOG.info("Newly registering node: {}", node.getId());
    } else {
      long regTime = pre.getRegistrationTime();
      // inherit registration time
      node.setRegistrationTime(regTime);
    }
```

**Review by xianjingfeng — APPROVED** [review]
LGTM

**Comment by rickyma** [comment]
Thanks for your contribution. @maobaolong 
Thanks for your reviews @xianjingfeng @advancedxy 
Merged to master.

**Comment by maobaolong** [comment]
@rickyma @advancedxy @xianjingfeng thanks for your review

**Pull request merged**